### PR TITLE
ci: use gha free runners

### PR DESCRIPTION
**Pull request by scotthart**
In order to accommodate the reduced disk size on the free runners, the sharding factor has been increased for the windows+cmake builds. With this additional sharding, all the windows+cmake builds are expected to pass and all the macos+cmake builds are expected to pass. bazel builds are likely going to fail until they can be similarly sharded.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290)

### CODE REVIEW

File: .github/workflows/macos-bazel.yml [hunk 1]
```diff
@@ -14,7 +14,7 @@ permissions:
 jobs:
   bazel:
     name: bazel + ${{ matrix.os }}
-    runs-on: ${{ matrix.os }}-large
+    runs-on: macos-latest
```
Comment (coryan) [inline .github/workflows/macos-bazel.yml:17]: `*-latest` can break the builds in unexpected ways, i.e., when GHA changes what specific version `latest` points to.  Better to use `macos-14` and then upgrade when ready to `macos-15` or whatever comes next.

File: .github/workflows/windows-bazel.yml [hunk 2]
```diff
@@ -14,9 +14,7 @@ permissions:
 jobs:
   bazel:
     name: bazel + ${{ matrix.msvc }} + ${{ matrix.compilation_mode }}
-    runs-on:
-      group: cpp-runners
-      labels: 'windows-2022'
+    runs-on: windows-latest
```
Comment (coryan) [inline .github/workflows/windows-bazel.yml:17]: Ditto, prefer `windows-2022` to avoid unexpected breaks.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[`.github/workflows/macos-bazel.yml` line 17 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290#gh-1622421099:-NzEOFVu60FZx7ylIU9g:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/002ddf68d7a5f89fa1292e65727cf479ca1aae83/.github/workflows/macos-bazel.yml#L17)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

`*-latest` can break the builds in unexpected ways, i.e., when GHA changes what specific version `latest` points to.  Better to use `macos-14` and then upgrade when ready to `macos-15` or whatever comes next.
</blockquote></details>

Done.
___
*[`.github/workflows/windows-bazel.yml` line 17 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290#gh-1622421768:-NzEOGOH8r451onaCAMV:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/002ddf68d7a5f89fa1292e65727cf479ca1aae83/.github/workflows/windows-bazel.yml#L17)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Ditto, prefer `windows-2022` to avoid unexpected breaks.
</blockquote></details>

Done.

File: .github/workflows/macos-bazel.yml [hunk 3]
```diff
@@ -14,15 +14,15 @@ permissions:
 jobs:
   bazel:
     name: bazel + ${{ matrix.os }}
-    runs-on: ${{ matrix.os }}-large
+    runs-on: macos-14
```
Comment (dbolduc) [inline .github/workflows/macos-bazel.yml:17]: leave as `${{ matrix.os }}` ?

File: .github/workflows/macos-cmake.yml [hunk 4]
```diff
@@ -18,7 +18,7 @@ permissions:
 jobs:
   cmake:
     name: cmake + ${{ matrix.os }} + ${{ matrix.shard }}
-    runs-on: ${{ matrix.os }}-xlarge
+    runs-on: macos-14
```
Comment (dbolduc) [inline .github/workflows/macos-cmake.yml:21]: same

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290)* status: 0 of 4 files reviewed, 4 unresolved discussions

___
*[`.github/workflows/macos-bazel.yml` line 17 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290#gh-1624544737:-NzTVOa53SMiBtAk3Jgv:b17iefa) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a7e71767a48c5d66d11e275dabf442645b461e96/.github/workflows/macos-bazel.yml#L17)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

leave as `${{ matrix.os }}` ?
</blockquote></details>

I'll send another PR to update.
___
*[`.github/workflows/macos-cmake.yml` line 21 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14290#gh-1624544925:-NzTVQEs8HsxW3dKC2rn:bb7od0r) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a7e71767a48c5d66d11e275dabf442645b461e96/.github/workflows/macos-cmake.yml#L21)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

same
</blockquote></details>

same

**Pull request merged**